### ISSUE TITLE
[feature request] Autotune TMA and non-TMA path for pointwise and reduction kernels

### ISSUE THREAD

**Issue by kshitij12345** [pytorch/pytorch#187175]
### Summary

When autotuning is enabled for GEMM kernels, both the TMA (Tensor Memory Accelerator) path and the non-TMA path are included as candidates and benchmarked against each other. This ensures the fastest path is always selected based on actual measured performance. However, for pointwise and reduction kernels, the TMA path is currently an all-or-nothing decision controlled by `config.triton.use_tensor_descriptor` — when it's enabled, the kernel is generated with TMA loads/stores and only TMA-compatible block sizes are considered. There is no autotuning between TMA and non-TMA variants.

### Motivation

For GEMM, the autotuner assembles candidates from both `mm_template` (non-TMA) and `persistent_tma_mm_template` (TMA) and benchmarks them in the same pool (see `torch/_inductor/kernel/mm.py`). This is valuable because TMA is not universally faster — its benefit depends on tensor shape, alignment, and access pattern.

For pointwise and reduction kernels, the same concern applies. Currently:

1. `TMACompatibilityChecker.can_use_tma()` decides at codegen time whether TMA is legal for each load/store.
2. `_maybe_filter_configs_for_tma_restrictions()` prunes block-size configs that violate TMA constraints.
3. But there's no mechanism to generate **both** a TMA variant and a non-TMA variant of the same pointwise/reduction kernel and let the autotuner pick the winner.

### Proposed Behavior (Two Phases)

**Phase 1 — While we are not yet confident in the TMA path for pointwise/reduction:**

Require both `use_tensor_descriptor=True` **and** autotuning (`max_autotune=True` or `max_autotune_pointwise=True`) to exercise TMA path for autotuning. When both are set, generate two kernel variants — one using TMA loads/stores and one using standard pointer arithmetic — and let the autotuner benchmark them head-to-head, just as it does for GEMM today. This gives us performance data and correctness coverage for the TMA path without committing to it unconditionally.

There are several recently filed issues that highlight the TMA path for pointwise/reduction is not yet fully mature:
- #186241
- #185513
- #185222
- #184714
- #184563

**Phase 2 — Once we are confident the TMA path is correct and mature:**

Enabling autotuning alone (without explicitly setting `use_tensor_descriptor`) should be sufficient to include TMA candidates alongside non-TMA candidates in the autotune pool — again mirroring how GEMM works today.

### Current Behavior

- **GEMM**: Both TMA and non-TMA templates are autotuned candidates → autotuner picks the best.
- **Pointwise/Reduction**: `config.triton.use_tensor_descriptor` is a binary switch. When `True`, only the TMA path is generated (with config filtering). When `False`, only the non-TMA path is generated. No head-to-head comparison.

### Additional context

Key code paths:
- GEMM autotune candidate assembly: `torch/_inductor/kernel/mm.py` (templates added to the same pool)
- Pointwise/reduction TMA eligibility: `TMACompatibilityChecker` in `torch/_inductor/codegen/triton.py`
- Config filtering for TMA: `_maybe_filter_configs_for_tma_restrictions()` in `torch/_inductor/runtime/triton_heuristics.py`
- TMA config: `torch/_inductor/config.py` → `triton.use_tensor_descriptor`

cc @ptrblck @msaroufim @eqy @tinglvv @nWEIdia @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by riccardofelluga** [comment]
Interested

**Comment by oonyshch** [comment]
Hi @kshitij12345 @riccardofelluga ! I'd like to help move this forward, as I currently have the bandwidth, and an arsenal of idle Hoppers and Blackwells to implement and test this :)) cc: @IvanYashchuk

Is there any work in progress I could build on, or would you be open to me taking this over / pairing on it? Wanted to check before duplicating effort.

Motivation-wise this lines up closely with recent Triton tensor-descriptor (TD/TMA) work I just did in vLLM, where making both operand loads go through tl.make_tensor_descriptor gave some significant perf wins, which is the same reason it's worth an autotuned candidate here rather than an all-or-nothing use_tensor_descriptor

So pls let me know if that's welcome, if I could help you and you'd prefer to split it up.

**Comment by jananisriram** [comment]
generally, aren't pointwise and reduction kernels memory-bound, in which case TMA is unlikely to be performant regardless? based on some benchmarks in https://github.com/pytorch/pytorch/pull/188823, it seems like device-side TMA is always slower than host-side TMA, and host-side TMA generally is slower than no TMA for pointwise/reduction.

**Comment by kshitij12345** [comment]
Running a benchmark on GB200 comparing Inductor TMA configurations for SiLU at production LLM shapes (LLama 4 intermediate_size=16384, DeepSeek-V3 intermediate_size=18432):

PyTorch version - 725887a
```
torch.compile SiLU (bf16), GB200, triton.testing.do_bench:

Shape              Label             no TMA   dev TMA  host TMA  dev/base  host/base
(4096, 4096)       small             44.9us    56.5us    51.8us     1.26x      1.15x
(8192, 8192)       medium            47.1us   111.7us    61.6us     2.37x      1.31x
(4096, 16384)      LLama4 bs=1       47.0us   111.7us    59.4us     2.38x      1.26x
(8192, 16384)      LLama4 bs=2       86.0us   155.9us    84.0us     1.81x      0.98x
(16384, 16384)     LLama4 bs=4      164.1us   298.6us   158.4us     1.82x      0.97x
(32768, 16384)     LLama4 bs=8      321.6us   584.4us   308.4us     1.82x      0.96x
(4096, 18432)      DSv3 bs=1         52.2us    92.6us    56.7us     1.77x      1.09x
(8192, 18432)      DSv3 bs=2         96.2us   174.1us    92.9us     1.81x      0.97x
(16384, 18432)     DSv3 bs=4        184.4us   334.2us   177.6us     1.81x      0.96x
(32768, 18432)     DSv3 bs=8        359.9us   654.5us   346.2us     1.82x      0.96x

no TMA    = use_tensor_descriptor=False (default)
dev TMA   = use_tensor_descriptor=True, enable_host_side_tma=False
host TMA  = use_tensor_descriptor=True, enable_host_side_tma=True
>1 = regression vs baseline
```

host-side TMA shows a small win (3–4%) at large production shapes (≥ 128M elements, i.e. LLama4 bs≥2, DSv3 bs≥2). At smaller shapes it regresses.

So, TMA should only be selected when the shape is large enough for the DMA engine's bandwidth advantage to pay off. A head-to-head autotune between TMA and non-TMA variants would naturally handle this shape-dependent crossover.

Repro

<details>

```python
"""
Benchmark torch.compile with TMA variants enabled vs baseline.

Tests all Inductor TMA configurations:
  1. torch.compile (default, no TMA)
  2. torch.compile + use_tensor_descriptor=True (device-side TMA)
  3. torch.compile + use_tensor_descriptor=True + enable_host_side_tma=True
"""

import torch
import torch._dynamo
import torch._inductor.config as inductor_config
import triton


def bench_compile_variants():
    device = torch.cuda.get_device_name(0)
    print(f"Device: {device}")
    print(f"Timing: triton.testing.do_bench (warmup=25, rep=100)")
    print()

    shapes = [
        (4096, 4096, "small"),
        (8192, 8192, "medium"),
        (4096, 16384, "LLama4 bs=1"),
        (8192, 16384, "LLama4 bs=2"),
        (16384, 16384, "LLama4 bs=4"),
        (32768, 16384, "LLama4 bs=8"),
        (4096, 18432, "DSv3 bs=1"),
        (8192, 18432, "DSv3 bs=2"),
        (16384, 18432, "DSv3 bs=4"),
        (32768, 18432, "DSv3 bs=8"),
    ]

    print(f"{'Shape':<18} {'Label':<14} {'no TMA':>9} {'dev TMA':>9} {'host TMA':>9} "
          f"{'dev/base':>9} {'host/base':>10}")
    print("-" * 85)

    for M, N, label in shapes:
        x = torch.randn(M, N, dtype=torch.bfloat16, device="cuda")

        # 1. Baseline: no TMA
        torch._dynamo.reset()
        with inductor_config.patch({
            "triton.use_tensor_descriptor": False,
        }):
            fn_base = torch.compile(
                lambda x: torch.nn.functional.silu(x),
                mode="max-autotune-no-cudagraphs",
                fullgraph=True,
            )
            for _ in range(3):
                fn_base(x)
            torch.cuda.synchronize()
            t_base = triton.testing.do_bench(lambda: fn_base(x), warmup=25, rep=100) * 1000

        # 2. Device-side TMA
        torch._dynamo.reset()
        t_dev = None
        try:
            with inductor_config.patch({
                "triton.use_tensor_descriptor": True,
                "triton.enable_host_side_tma": False,
                "assume_aligned_inputs": True,
            }):
                fn_dev = torch.compile(
                    lambda x: torch.nn.functional.silu(x),
                    mode="max-autotune-no-cudagraphs",
                    fullgraph=True,
                )
                for _ in range(3):
                    fn_dev(x)
                torch.cuda.synchronize()
                t_dev = triton.testing.do_bench(lambda: fn_dev(x), warmup=25, rep=100) * 1000
        except Exception as e:
            pass

        # 3. Host-side TMA
        torch._dynamo.reset()
        t_host = None
        try:
            with inductor_config.patch({
                "triton.use_tensor_descriptor": True,
                "triton.enable_host_side_tma": True,
                "assume_aligned_inputs": True,
            }):
                fn_host = torch.compile(
                    lambda x: torch.nn.functional.silu(x),
                    mode="max-autotune-no-cudagraphs",
                    fullgraph=True,
                )
                for _ in range(3):
                    fn_host(x)
                torch.cuda.synchronize()
                t_host = triton.testing.do_bench(lambda: fn_host(x), warmup=25, rep=100) * 1000
        except Exception as e:
            pass

        # Format results
        dev_str = f"{t_dev:>7.1f}us" if t_dev else "   FAIL"
        host_str = f"{t_host:>7.1f}us" if t_host else "   FAIL"
        dev_ratio = f"{t_dev / t_base:>8.2f}x" if t_dev else "     N/A"
        host_ratio = f"{t_host / t_base:>9.2f}x" if t_host else "      N/A"

        shape_str = f"({M}, {N})"
        print(f"{shape_str:<18} {label:<14} {t_base:>7.1f}us {dev_str} {host_str} "
              f"{dev_ratio} {host_ratio}")

    print()
    print("Legend:")
    print("  no TMA    = torch.compile default (use_tensor_descriptor=False)")
    print("  dev TMA   = use_tensor_descriptor=True, enable_host_side_tma=False")
    print("  host TMA  = use_tensor_descriptor=True, enable_host_side_tma=True")
    print("  dev/base  = device-side TMA relative to baseline (>1 = regression)")
    print("  host/base = host-side TMA relative to baseline (>1 = regression)")


if __name__ == "__main__":
    bench_compile_variants()
```

</details>

**Comment by riccardofelluga** [comment]
Hi @oonyshch, apologies for the long delay in the reply. Sure! I'd be happy to get any help on this, are you still interested/ do you still have bandwidth?

**Comment by oonyshch** [comment]
> Hi [@oonyshch](https://github.com/oonyshch), apologies for the long delay in the reply. Sure! I'd be happy to get any help on this, are you still interested/ do you still have bandwidth?

Hi @riccardofelluga, of course! What would you like me to do?
